components/selectors/matching.rs

```diff
@@ -112,12 +112,56 @@ pub fn matches<E>(selector_list: &[Selector<E::Impl>],
     })
 }
 
+fn may_match<E>(mut selector: &ComplexSelector<E::Impl>,
+                bf: &BloomFilter)
+                -> bool
+    where E: Element,
+{
+    // See if the bloom filter can exclude any of the descendant selectors, and
+    // reject if we can.
+    loop {
+         match selector.next {
+             None => break,
+             Some((ref cs, Combinator::Descendant)) => selector = &**cs,
+             Some((ref cs, _)) => {
+                 selector = &**cs;
+                 continue;
+             }
+         };
```

Comment:
Still I haven't got all the details, but look like this prevents checking the first selector(the rightmost one) in the bloom filter. Maybe move this match to the end of the loop?